io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,394 @@
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*
+*/
+
+#include <pcl/io/eigen.h>
+#include <pcl/io/boost.h>
+#include <pcl/io/grabber.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+
+#include <librealsense2/rs.hpp>
+#include <pcl/io/real_sense_2_grabber.h>
+#include <pcl/common/time.h>
+
+namespace pcl
+{
+  RealSense2Grabber::RealSense2Grabber (const std::string& file_name_or_serial_number)
+    : running_ (false)
+    , quit_ (false)
+    , signal_PointXYZ (nullptr)
+    , signal_PointXYZI (nullptr)
+    , signal_PointXYZRGB (nullptr)
+    , signal_PointXYZRGBA (nullptr)
+    , fps_ (0)
+    , device_width_ (424)
+    , device_height_ (240)
+    , target_fps_ (30)
+    , file_name_or_serial_number_(file_name_or_serial_number)
+  {
+    signal_PointXYZ = createSignal<signal_librealsense_PointXYZ> ();
+    signal_PointXYZI = createSignal<signal_librealsense_PointXYZI> ();
+    signal_PointXYZRGB = createSignal<signal_librealsense_PointXYZRGB> ();
+    signal_PointXYZRGBA = createSignal<signal_librealsense_PointXYZRGBA> ();
+  }
+
+  RealSense2Grabber::~RealSense2Grabber ()
+  {
+    stop ();
+
+    disconnect_all_slots<signal_librealsense_PointXYZ> ();
+    disconnect_all_slots<signal_librealsense_PointXYZI> ();
+    disconnect_all_slots<signal_librealsense_PointXYZRGB> ();
+    disconnect_all_slots<signal_librealsense_PointXYZRGBA> ();
+
+    thread_.join ();
+
+    pipe_.stop ();
+  }
+
+  void 
+  RealSense2Grabber::start ()
+  {
+    if (isRunning ())
+      return;
+
+    running_ = true;
+
+    rs2::config cfg;
+
+    // capture from file
+    if (file_name_or_serial_number_.rfind (".bag") == file_name_or_serial_number_.length() - 4)
+    {
+      cfg.enable_device_from_file (file_name_or_serial_number_);
+    }
+    else
+    {
+      if (!file_name_or_serial_number_.empty ())
+        cfg.enable_device (file_name_or_serial_number_);
+
+      cfg.enable_stream (RS2_STREAM_COLOR, device_width_, device_height_, RS2_FORMAT_RGB8, target_fps_);
+      cfg.enable_stream (RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_Z16, target_fps_);
+      cfg.enable_stream (RS2_STREAM_INFRARED, device_width_, device_height_, RS2_FORMAT_Y8, target_fps_);
+    }    
+
+    rs2::pipeline_profile prof = pipe_.start (cfg);
+
+    assert (prof.get_stream (RS2_STREAM_COLOR).format () == RS2_FORMAT_RGB8);
+    assert (prof.get_stream (RS2_STREAM_DEPTH).format () == RS2_FORMAT_Z16);
+
+    ir_format_ = prof.get_stream (RS2_STREAM_INFRARED).format (); 
+
+    thread_ = std::thread (&RealSense2Grabber::threadFunction, this);
+
+  }
+
+  void 
+  RealSense2Grabber::stop ()
+  {
+    std::lock_guard<std::mutex> guard (mutex_);
+
+    quit_ = true;
+    running_ = false;
+  }
+
+  bool 
+  RealSense2Grabber::isRunning () const
+  {
+    std::lock_guard<std::mutex> guard (mutex_);
```

Comment:
This one is moderately troublesome because `isRunning` feels like it should be a super quick function call, but it is waiting on this entire [block](https://github.com/PointCloudLibrary/pcl/pull/2214/files#diff-2c7a8b41ca6e6bcff66df4b94d4d76b0R345) to finish processing. I would remove this lock guard and reverse the order of lines 125 and 126.